typescript_test/test.ts

```diff
@@ -0,0 +1,346 @@
+  connected({bar: 42, baz: 123});
+}
+
+function testInvalidTypeInCombinator() {
+  // typings:expect-error
+  createSelector(
+    (state: {foo: string}) => state.foo,
+    (foo: number) => foo,
+  );
+
+  // typings:expect-error
+  createSelector(
+    (state: {foo: string, bar: number, baz: boolean}) => state.foo,
+    state => state.bar,
+    state => state.baz,
+    (foo: string, bar: number, baz: boolean, fizz: string) => {}
+  );
+}
+
+function testRestParameters() {
+  createSelector(
+    (state: {foo: string, bar: number}) => state.foo,
+    state => state.foo,
+    state => state.foo,
+    state => state.foo,
+    state => state.foo,
+    state => state.bar,
+    state => state.bar,
+    state => state.bar,
+    state => state.bar,
+    state => state.bar,
+    (foo1: string, foo2: string, foo3: string, foo4: string, foo5: string,
+     bar1: number, bar2: number, bar3: number, bar4: number, bar5: string) => {
+    }
+  );
+}
+
+function testParametricSelector() {
+  type State = {foo: string;};
+  type Props = {bar: number};
+
+  const selector = createSelector(
+    (state: State) => state.foo,
+    (state, props: Props) => props.bar,
+    (foo, bar) => ({foo, bar}),
+  );
+
+  // typings:expect-error
+  selector({foo: 'fizz'});
+  // typings:expect-error
+  selector({foo: 'fizz'}, {bar: 'baz'});
+
+  const ret = selector({foo: 'fizz'}, {bar: 42});
+  const foo: string = ret.foo;
+  const bar: number = ret.bar;
+
+  const selector2 = createSelector(
+    (state) => state.foo,
+    (state) => state.foo,
+    (state) => state.foo,
+    (state) => state.foo,
+    (state) => state.foo,
+    (state: State, props: Props) => props.bar,
+    (foo1, foo2, foo3, foo4, foo5, bar) => ({
+      foo1, foo2, foo3, foo4, foo5, bar,
+    }),
+  );
+
+  selector2({foo: 'fizz'}, {bar: 42});
+}
+
+function testArrayArgument() {
+  const selector = createSelector([
+    (state: {foo: string}) => state.foo,
+    state => state.foo,
+    (state, props: {bar: number}) => props.bar,
+  ], (foo1, foo2, bar) => ({foo1, foo2, bar}));
+
+  const ret = selector({foo: 'fizz'}, {bar: 42});
+  const foo1: string = ret.foo1;
+  const foo2: string = ret.foo2;
+  const bar: number = ret.bar;
+
+  // typings:expect-error
+  createSelector([
+    (state: {foo: string}) => state.foo,
+  ]);
+
+  // typings:expect-error
+  createSelector([
+    (state: {foo: string}) => state.foo,
+    (state: {bar: number}) => state.bar,
+  ], (foo, bar) => {});
+
+  // typings:expect-error
+  createSelector([
+    (state: {foo: string}) => state.foo,
+    state => state.foo,
+  ], (foo: string, bar: number) => {});
+
+  createSelector([
+    (state: {foo: string}) => state.foo,
+    state => state.foo,
+    state => state.foo,
+    state => state.foo,
+    state => state.foo,
+    state => state.foo,
+    state => state.foo,
+    state => state.foo,
+    state => state.foo,
+    state => state.foo,
+  ], (foo1, foo2, foo3, foo4, foo5, foo6, foo7, foo8, foo9, foo10) => {
```

Comment:
Yep, apparently return types of input selectors are not inferred and this hack doesn't work:
```ts
export function createSelector<S, P, R1, ..., R8, SS extends [
  ParametricSelector<S, P, R1>,
  ...,
  ParametricSelector<S, P, R8>
] & {[index: number]: ParametricSelector<S, P, any>}, T>(
  selectors: SS,
  combiner: (res1: R1, ..., res8: R8, ...rest: any[]) => T,
): ParametricSelector<S, P, T>;
```
I guess we'll have to stick with
```ts
export function createSelector<S, P, T>(
  selectors: ParametricSelector<S, P, any>[],
  combiner: (...rest: any[]) => T,
): ParametricSelector<S, P, T>;
```